### PR TITLE
Fix #4543 - E+ 22.1.0: Wrap SetpointManager:SystemNodeReset:Temperature and SetpointManager:SystemNodeReset:Humidity

### LINKED ISSUE #4543: E+ 22.1.0: Wrap SetpointManager:SystemNodeReset:Temperature and SetpointManager:SystemNodeReset:Humidity
Enhancement Request
--------------

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fix #4543
 - Companion PR: https://github.com/NREL/OpenStudio-resources/pull/170

### Pull Request Author

 - [x] Model API Changes / Additions
 - [x] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [x] Model API methods are tested (in `src/model/test`)
 - [x] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [x] If a new object or method, added a test in NREL/OpenStudio-resources: https://github.com/NREL/OpenStudio-resources/pull/170
 - [x] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [x] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [x] All new and existing tests passes
 - [x] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/model/SetpointManagerSystemNodeResetHumidity.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                                                                                             Model_Impl* model, bool keepHandle)
+      : SetpointManager_Impl(other, model, keepHandle) {
+      OS_ASSERT(other.iddObject().type() == SetpointManagerSystemNodeResetHumidity::iddObjectType());
+    }
+
+    SetpointManagerSystemNodeResetHumidity_Impl::SetpointManagerSystemNodeResetHumidity_Impl(const SetpointManagerSystemNodeResetHumidity_Impl& other,
+                                                                                             Model_Impl* model, bool keepHandle)
+      : SetpointManager_Impl(other, model, keepHandle) {}
+
+    const std::vector<std::string>& SetpointManagerSystemNodeResetHumidity_Impl::outputVariableNames() const {
```
Comment (joseph-robertson) [inline src/model/SetpointManagerSystemNodeResetHumidity.cpp:65]: Need to override this if it's empty anyway?

**Review by joseph-robertson — COMMENTED** [review]
Looks good. I mostly just have a few minor questions/thoughts. The two objects look very similar, so assume that any questions for the first also apply to the second.

File: src/energyplus/ReverseTranslator.cpp [hunk 2]
```diff
@@ -810,6 +810,14 @@ namespace energyplus {
         //modelObject = translateSetpointManagerSingleZoneReheat(workspaceObject);
         break;
       }
+      case openstudio::IddObjectType::SetpointManager_SystemNodeReset_Humidity: {
+        //modelObject = translateSetpointManagerSystemNodeResetHumidity(workspaceObject);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ReverseTranslator.cpp:814]: Until we invest into the effort of ReverseTranslating Loops (which we don't currently), I must leave it disabled yes

File: src/model/SetpointManagerSystemNodeResetHumidity.cpp [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    void SetpointManagerSystemNodeResetHumidity_Impl::resetSetpointNode() {
+      bool result = setString(OS_SetpointManager_SystemNodeReset_HumidityFields::SetpointNodeorNodeListName, "");
+      OS_ASSERT(result);
+    }
+
+  }  // namespace detail
+
+  SetpointManagerSystemNodeResetHumidity::SetpointManagerSystemNodeResetHumidity(const Model& model)
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/SetpointManagerSystemNodeResetHumidity.cpp:182]: No. I specifically made IDD changes to make the nodes not required-field. This is the way all SetpointManagers behave.

File: src/model/SetpointManagerSystemNodeResetHumidity.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    : SetpointManager(SetpointManagerSystemNodeResetHumidity::iddObjectType(), model) {
+    OS_ASSERT(getImpl<detail::SetpointManagerSystemNodeResetHumidity_Impl>());
+
+    // From E+ 5ZoneSystemNodeReset.idf (22.1.0)
+    bool ok = setControlVariable("MaximumHumidityRatio");
+    OS_ASSERT(ok);
+    setSetpointatLowReferenceHumidityRatio(0.00924);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/SetpointManagerSystemNodeResetHumidity.cpp:189]: Yeah I don't think that's needed here. I'm setting positive, non zero, numeric values. That's never going to change. The Choice field could change and I could forget to update the Ctor though.

File: src/model/SetpointManagerSystemNodeResetHumidity.cpp (hunk 1 above)
Comment (jmarrec) [reply src/model/SetpointManagerSystemNodeResetHumidity.cpp:65]: You do. This is a pure virtual method in ModelObject, so you must define it.

https://github.com/NREL/OpenStudio/blob/2924b53962e70d425a77f677dfcf74de1f1458af/src/model/ModelObject_Impl.hpp#L132

File: src/energyplus/ForwardTranslator/ForwardTranslateSetpointManagerSystemNodeResetHumidity.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    // Low Reference Humidity Ratio: Required Double
+    double lowReferenceHumidityRatio = modelObject.lowReferenceHumidityRatio();
+    idfObject.setDouble(SetpointManager_SystemNodeReset_HumidityFields::LowReferenceHumidityRatio, lowReferenceHumidityRatio);
+
+    // High Reference Humidity Ratio: Required Double
+    double highReferenceHumidityRatio = modelObject.highReferenceHumidityRatio();
+    idfObject.setDouble(SetpointManager_SystemNodeReset_HumidityFields::HighReferenceHumidityRatio, highReferenceHumidityRatio);
+
+    // Reference Node Name: Required Node
+    if (auto node = modelObject.referenceNode()) {
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateSetpointManagerSystemNodeResetHumidity.cpp:79]: That's true. I suppose I could remove the check.

File: src/energyplus/ForwardTranslator/ForwardTranslateSetpointManagerSystemNodeResetHumidity.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  boost::optional<IdfObject>
+    ForwardTranslator::translateSetpointManagerSystemNodeResetHumidity(model::SetpointManagerSystemNodeResetHumidity& modelObject) {
+
+    // Translation is triggered via the setpointNode already, so no need to check
+    if (!modelObject.referenceNode()) {
+      LOG(Error, modelObject.briefDescription() << " does not have a Reference Node attached, it will be not be translated.");
+      return boost::none;
+    }
+
+    // Instantiate an IdfObject of the class to store the values,
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateSetpointManagerSystemNodeResetHumidity.cpp]: Done in daa536984c9fb2678bf6e0390ca7af4d699a68ed

File: src/energyplus/ForwardTranslator/ForwardTranslateSetpointManagerSystemNodeResetHumidity.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateSetpointManagerSystemNodeResetHumidity.cpp]: Done in daa536984c9fb2678bf6e0390ca7af4d699a68ed

**Pull request merged**